docs(api): refactor-tip-handling (#13383)
OpenTrons/opentrons-api: api/docs/v2/new_atomic_commands.rst

    pipette.pick_up_tip()  # picks up tip from rack location B1
    pipette.drop_tip()     # drops tip in trash bin 

> If you omit the ``tip_rack`` argument from the ``pipette`` variable, the API will throw an error. You must pass in the tip rack's location to ``pick_up_tip`` like this::

Reviewer: we always `raise` instead of `throw` errors, because 🐍

                pipette = protocol.load_instrument(
                    instrument_name="p300_single_gen2",
                    mount="left",
                    tip_racks=[tip_rack_1],
                )  
                # Put other protocol commands here

Manipulating Pipette Tips
=========================

> Your robot needs to attach a disposable tip to the pipette before it can aspirate or dispense liquids. The API provides three basic functions that help the robot attach and manage pipette tips during a protocol run. These methods are :py:meth:`.InstrumentContext.pick_up_tip`, :py:meth:`.InstrumentContext.drop_tip`, and :py:meth:`.InstrumentContext.return_tip`. Respectively, these methods tell the robot to pick up a tip from a tip rack, drop a tip into the trash (or another location), and return a tip to its location in the tip rack. Let's look at examples for each of these functions. Each code snippet works with the sample code shown above.

Reviewer: 👍 
you can link to 
```
:ref:`protocol-template`
```
even though it'll just show up as plain text for now.

Reviewer: there may also be some Py-pedantry here about `function` vs `method`

Picking Up a Tip
----------------

To pick up a tip, call the :py:meth:`~.InstrumentContext.pick_up_tip` method without any arguments::
    
    pipette.pick_up_tip()

This simple statement works because the variable ``tiprack_1`` in the sample protocol includes the on-deck location of the tip rack (Flex ``location="D3"``, OT-2 ``location=3``) *and* the ``pipette`` variable includes the argument ``tip_racks=[tiprack_1]``. Given this information, the robot moves to the tip rack and picks up a tip from position A1 in the rack. On subsequent calls to ``pick_up_tip()``, the robot will use the next available tip. For example::

    pipette.pick_up_tip()  # picks up tip from rack location A1
    pipette.drop_tip()     # drops tip in trash bin
    pipette.pick_up_tip()  # picks up tip from rack location B1
    pipette.drop_tip()     # drops tip in trash bin 

If you omit the ``tip_rack`` argument from the ``pipette`` variable, then you must pass in the tip rack's location to ``pick_up_tip`` like this::
    
    pipette.pick_up_tip(tiprack_1['A1'])
    pipette.drop_tip()
    pipette.pick_up_tip(tiprack_1['B1'])

> However, coding the location of each tip is inefficient. Let's use a ``for`` loop to automate a sequential tip pick up process.

Reviewer: I would mention here that if you call `pick_up_tip()` without associating tip racks with the pipette, the API will raise an error.

Then this sentence could go into the next section, and be reframed as 
```
Coding the location of each tip is inefficient. It's usually best to let the API 
track pickup location for you. 
```
_next subsection_
```
For example, let's use a ``for`` loop to automate picking up and dropping 
a large number of tips from a rack.
```

Author: Adding text about API errors:

> If you omit the ``tip_rack`` argument from the ``pipette`` variable, the API will throw an error. You must pass in the tip rack's location to ``pick_up_tip`` like this::

Conclusion to this subsection revised too:

> If coding the location of each tip seems inefficient or tedious, try using a ``for`` loop to automate a sequential tip pick up process. When using a loop, the API tracks and manages the tip pickup locations automatically. But ``pick_up_tip`` is still a powerful feature. It lets you maintain control over tip use when that’s important in your protocol.

Leads in to the section on `for` loops.

.. versionadded:: 2.0

> .. needs better title?

Reviewer: `Automating Pickup`?

Author: Using that.

Loops and Tip Pick Up
---------------------

> A ``for`` loop and Python's :py:class:`range` class gives you a better way to automate the tip pickup process. It eliminates the need to call ``pick_up_tip()`` multiple times. For example, this snippet tells the robot to sequentially use all the tips in a 96-tip rack::

Reviewer: This makes it sound a little like you should _always_ use loops with `pick_up_tip` or that it's some kind of best practice. A loop works great with `pick_up_tip` when it meets the needs of the protocol. It only suits our needs here because we want to quickly show burning through a whole rack of tips.

I basically don't want to discourage users from calling `pick_up_tip()` sans argument piecemeal throughout their protocol. It's meant for that too.

Author: Made changes to try and reduce this impression.

    for i in range(96):
        pipette.pick_up_tip()

>         pipette.drop_tip()

Reviewer: could add a comment line here
```suggestion
        # liquid handling commands
        pipette.drop_tip()
```
…to make it not seem like we're just throwing a way a whole tip rack for no reason

> If your protocol requires a lot of tips, add a second tip rack to the protocol, and sum the tip count in the range. The robot will loop through both racks. For example, let's add another tip rack to the sample protocol::

Reviewer: "and sum the tip count in the range" feels out of order, since it doesn't come until a couple snippets later. maybe
```suggestion
If your protocol requires a lot of tips, add a second tip rack to the protocol. Then associate it with your pipette and increase the number of repetitions in the loop. The robot will loop through both racks. 

First, add another tip rack to the sample protocol::
```

    tiprack_2 = protocol.load_labware(
        load_name="opentrons_flex_96_tiprack_1000ul",
        location="C3"
    )

Next, revise the pipette's ``load_instrument()`` method to include the new tip rack in the ``tip_rack`` argument::

    pipette = protocol.load_instrument(
        instrument_name="flex_1channel_1000",
        mount="left",

>         tip_racks=[tiprack_1, tip_rack_2],

Reviewer: ```suggestion
        tip_racks=[tiprack_1, tiprack_2],
```

    ) 

Finally, sum the tip count in the range::

    for i in range(192):
        pipette.pick_up_tip()
        pipette.drop_tip()

For a more advanced "real-world" example, take a moment to review the :ref:`off-deck location protocol <off-deck-location>` on the :ref:`moving-labware` page. This example also uses a ``for`` loop to iterate through a tip rack, but it includes other commands that pause the protocol and let you replace an on-deck tip rack with another rack stored in an off-deck location.

Dropping a Tip
--------------

To drop a tip in the trash bin, call the :py:meth:`~.InstrumentContext.drop_tip` method without any arguments::
    
    pipette.pick_up_tip()

You can also specify where to drop the tip by passing in a location. For example, this code drops a tip in the trash bin and returns another tip to the tip rack::

    pipette.pick_up_tip()            # picks up tip from rack location A1
    pipette.drop_tip()               # drops tip in trash bin 
    pipette.pick_up_tip()            # picks up tip from rack location B1

>     pipette.drop_tip(tiprack['B1'])  # drops tip in rack location B1

Reviewer: i'm not sure why you would use `drop_tip` to put a tip back where it came from, instead of `return_tip`. maybe change to `A1` here to show that it's possible to move a tip to a _different_ well of the rack.

Reviewer: it may also be useful to see whether you can drop tips in arbitrary labware. like can i make a bonus trash container out of a 1-well reservoir? i forget what limits the API has there.

Author: > i'm not sure why you would use drop_tip to put a tip back where it came from, instead of return_tip. maybe change to A1 here to show that it's possible to move a tip to a different well of the rack.

I changed the text for `drop_tip` to well A1 to demonstrate returning it to a previously used well.

> it may also be useful to see whether you can drop tips in arbitrary labware. like can i make a bonus trash container out of a 1-well reservoir? i forget what limits the API has there.

I tried with an Agilent 1 well reservoir (290 ml) and an Axigen reservoir (90 ml). Validation returns a "Labware . . . has no tip length defined" error. Not sure how to add `tip_length` to the sample code. Will ask, but to keep things going, skipping for now. I get the potential use case though. Maybe you'd want to use a reservoir instead of contaminating the trash bin -- something like that.

Reviewer: oh hey look what i found: RSS-277

Author: This drove me a little crazy at the end of yesterday. I was trying all sorts of things.

`pipette.drop_tip(reservoir.wells_by_name()['A1'])`

setting weird variables:

```
# test alt trash
alt_trash_reservoir = reservoir
# and then
pipette.drop_tip(alt_trash_reservoir)
```

And even enlisted Jeremy in the struggle.

.. versionadded:: 2.0

.. _pipette-return-tip:

Return Tip
===========

> To return a tip to its original location, call the :py:meth:`~.InstrumentContext.return_tip` method without any arguments::

Reviewer: maybe worth noting that it _never_ takes any position argument

Author: Slight text revision:

> To return a tip to its original location, call the ... method with no arguments::

Also, the text provides a link to the method. I'm gonna say let 'em click the link and check the original to see that.

    pipette.return_tip()

> .. this section below was a long note

Reviewer: ```
..
   fun fact
   you can make a multi-line RST comment like this

   it's all a comment

now this isn't a comment
```

Author: Thank you. I wasn't sure how to do this. Will try it.

.. a lot of info for a call-out
.. let's try it in a section
.. important info, let's give it some attention

Working With Used Tips
----------------------

API versions 2.2 or higher consider tips as "used" after being picked up. For example, if the robot picked up a tip from rack location A1 and then returned it to the same location, it will not attempt to pick up this tip again, unless explicitly specified. Instead, the robot will pick up a tip starting from rack location B1. For example::

    pipette.pick_up_tip()                # picks up tip from rack location A1
    pipette.return_tip()                 # drops tip in rack location A1
    pipette.pick_up_tip()                # picks up tip from rack location B1
    pipette.drop_tip()                   # drops tip in trash bin

>     pipette.pick_up_tip(tiprack_1['A1']) # picks up tip from rack location A1

Reviewer: add a `.. versionchanged: 2.2` here

Author: Added version changed indicator.

.. can this be removed, is it helpful?

> Also in API Version 2.2, the return tip height was corrected to utilize values determined by hardware testing. This is more in-line with return tip behavior from Python Protocol API Version 1.

Reviewer: this is ancient history, scrap it imo

Author: Removed.

> API versions 2.0 and 2.1 treated returned tips as unused items. They could be picked up again without an explicit argument. For example:: 

Reviewer: A slight reframe possibility here.

At the top of the section, instead of talking about versions, begin `The API considers tips...`. Then include the version changed notice I flagged above. Then put this sentence and the snippet below inside a note.

Author: I don't quite follow this suggestion. If the objection is to mentioning specific versions, lets try this:

Starting sentence:

> Currently, the API considers tips as "used" after being picked up.

Next paragraph about API v2.0, 2.1:

> Early API versions treated returned tips as unused items.

Also, discussing how the API treats tips as "used" seems like good and interesting info. I'd suggest a Note isn't quite right and recommend keeping this in its own section.

Subsection ends with:

> .. versionchanged: 2.2

    pipette.pick_up_tip()  # picks up tip from rack location A1
    pipette.return_tip()   # drops tip in rack location A1
    pipette.pick_up_tip()  # picks up tip from rack location A1

> 

Reviewer: We may also want a subsection explaining `InstrumentContext.has_tip` — it's useful for avoiding "hey you've already got a tip" errors with `pick_up_tip()`. And we should state somewhere that it's purely based on API-internal state. OT-2 pipettes have no way of knowing whether a tip is attached. And I don't think the Flex sensors have been wired up to `has_tip`. So it really just keeps track of whether the API _thinks_ the pipette _should_ have a tip attached based on previous instructions.

Author: Isn't there an optical sensor on the Flex pipette that knows if a tip is attached? I think it's the OT-2 that doesn't know about tips in the same way as a Flex does. So is it the OT-2 that only knows if `has_tip` is yes or no based on previous commands?  I'm not sure what to do here.

Reviewer: Ticketed as RTC-334 so we can revisit it later.
